filters/apiusagemonitoring/filter.go

```diff
@@ -0,0 +1,126 @@
+package apiusagemonitoring
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/zalando/skipper/filters"
+	"net/http"
+	"time"
+)
+
+const (
+	metricCountAll  = "http_count"
+	metricCount500s = "http5xx_count"
+	metricCount400s = "http4xx_count"
+	metricCount300s = "http3xx_count"
+	metricCount200s = "http2xx_count"
+	metricCount100s = "http1xx_count"
+	metricLatency   = "latency"
+)
+
+const (
+	stateBagKeyPrefix = "filter." + Name + "."
+	stateBagKeyState  = stateBagKeyPrefix + "state"
+)
+
+// apiUsageMonitoringFilter implements filters.Filter interface and is the structure
+// created for every route invocation of the `apiUsageMonitoring` filter.
+type apiUsageMonitoringFilter struct {
+	Paths []*pathInfo
+}
+
+// apiUsageMonitoringFilterContext contains information about the metrics tracking
+// for one HTTP exchange (one routing). It serves to pass information from
+// the `Request` call to the `Response` call (stored in the context's `StateBag`).
+type apiUsageMonitoringFilterContext struct {
+	// Begin is the earliest point in time where the request is observed
+	Begin time.Time
+}
+
+func (f *apiUsageMonitoringFilter) Request(c filters.FilterContext) {
+	// Gathering information from the initial request for further metrics calculation
+	c.StateBag()[stateBagKeyState] = &apiUsageMonitoringFilterContext{
+		Begin: time.Now(),
+	}
+}
+
+func (f *apiUsageMonitoringFilter) Response(c filters.FilterContext) {
+	mfc, ok := c.StateBag()[stateBagKeyState].(*apiUsageMonitoringFilterContext)
+	if !ok {
+		log.Debugf("Call not tracked (key %q not found in StateBag)", stateBagKeyState)
+		return
+	}
+
+	request, response, metrics := c.Request(), c.Response(), c.Metrics()
+	metricsName := f.getMetricsName(request)
+
+	// METRIC: Count
+	metrics.IncCounter(metricsName.CountAll)
+
+	// METRIC: Response Status Range Count
+	classMetricsIndex := (response.StatusCode / 100) - 1
+	if classMetricsIndex < 0 || classMetricsIndex >= 5 {
+		log.Warnf(
+			"Response HTTP Status Code %d is invalid. Response status code metric not tracked for this call.",
+			response.StatusCode)
+	} else {
+		metrics.IncCounter(metricsName.CountPerStatusCodeRange[classMetricsIndex])
+	}
+
+	// METRIC: Latency
+	metrics.MeasureSince(metricsName.Latency, mfc.Begin)
+
+	log.Debugf("Pushed metrics prefixed by %q", metricsName.GlobalPrefix)
+}
+
+// String returns a JSON representation of the filter prefixed by its type.
+func (f *apiUsageMonitoringFilter) String() string {
+	var js string
+	if jsBytes, err := json.Marshal(f); err == nil {
+		js = string(jsBytes)
+	} else {
+		js = fmt.Sprintf("<%s>", err)
+	}
+	return fmt.Sprintf("%T %s", f, js)
+}
+
+// getMetricsName returns the structure with names of the metrics for this specific context.
+// If it is not already cached, it is generated and cached to speed up next calls.
+func (f *apiUsageMonitoringFilter) getMetricsName(req *http.Request) *specificMetricsName {
```

Comment:
Good function, but it should be named `func ... metricNames(...) *metricNames` to make clear, that multiple metric names are returned. I think there is no need to have a `specific` in front of the struct.

Reply:
This is an English grammar question.

Metric Names

or

Metrics Name

Where goes the `s`?